docs/dev/data-identity.rst

```diff
@@ -0,0 +1,250 @@
+.. _data_identity:
+
+###########################
+DFHack Data Identity System
+###########################
+
+This article is an attempt to describe DFHack's data identity system.
+DFHack internally has a collection of C++ classes that provide metedata about the data in DF itself as well as data used
+by various components of DFHack. This metadata is used primarily to enable the Lua scripting system to access data
+held by Dwarf Fortress in a transparent manner, but is also used for several other purposes within DFHack.
+
+The base class of the identity system is the class ``type_identity``, defined in :source:`DataDefs.h <library/include/DataDefs.h>`. A ``type_identity`` object
+provides information about one *type* of data object, in either Dwarf Fortress or DFHack, that can be manipulated as a discrete entity in Lua.
+With one specific exception (``global_identity``), there is a one-to-one relationship between C++ types and ``type_identity`` objects.
+In Lua, objects that are being managed via the data identity system are represented as a Lua userdata object. The userdata object
+contains both a pointer to the C++ object itself and a pointer to a ``type_identity`` object that describes the data pointed
+by that pointer. Note that the userdata object does not own the objects pointed to by these pointers, and the Lua engine is
+never responsible for managing their lifetimes.
+
+``type_identity`` defines the following public methods:
+
+- ``byte_size``: returns the size, in bytes, of the object held
+
+- ``type``: returns an enum (of type ``identity_type``) classifying the object held.
+
+- ``getFullName``: returns a string that describes the type. This will usually be similar to a C++ ``typedef``, although this is not guaranteed.
+
+- ``lua_read``: Used by the Lua engine to "read" the data from a C++ object into the Lua state.
+
+- ``lua_write``: Used by the Lua engine to "write" a value from the Lua state into a C++ data object.
+
+- ``build_metatable``: Create a Lua metatable in the specified Lua state corresponding to this type identity.
+
+- ``is_primitive``: indicates that ``lua_read`` will store a *copy* of the object on the Lua stack instead of a non-owning reference to it. Used for types that have direct representations in Lua: numbers, booleans, simple strings
+
+- ``is_constructed``: indicates that creating a C++ instance of this type requires the use of a possibly nontrivial constructor. A type identity that is both primitive and constructed cannot be inserted into a container. At the moment the only type identity that is both primitive and constructed is ``stl_string_identity``, which wraps the C++ ``std::string`` type.
```

Comment:
I find this slightly puzzling, because I have definitely seen ` map<string, map<string, string>>` being returned to Lua. https://github.com/DFHack/dfhack/blob/703123e5ec5f2c8d995dbeb9da053b11e37bd5d4/plugins/buildingplan/buildingplan.cpp#L974
Maybe this needs some clarification?

Reply:
when an object like this is returned to lua, the object on the lua stack will be a nonowning reference. which suggests that that object might well leak if/when lua GCs the userdata object. we may need to look into that

Reply:
update: that particular instance is handled _outside_ of the identity trait system; it's handled by `Lua::Push<T>`, which does not always do the same thing as `identity_traits<T>::get->lua_read`. the former handles some special cases that the latter does not. this may or may not be desirable.

Reply:
Yes, I saw the thread on discord. In light of that, this is probably fine.